src/selectors/mergeThoughts.ts

```diff
@@ -0,0 +1,155 @@
+import { omit } from 'lodash'
+import { getThoughtById } from '.'
+import { Index, Lexeme, Parent, Path, PendingMerge, State } from '../@types'
+import { appendToPath, hashThought, head, normalizeThought, timestamp } from '../util'
+import { getSessionId } from '../util/sessionManager'
+import { getNextRankById } from './getNextRank'
+import move from './move'
+
+interface RecursiveMoveData {
+  thoughtIndexUpdates: Index<Lexeme | null>
+  contextIndexUpdates: Index<Parent | null>
+  pendingMerges: PendingMerge[]
+}
+
+/**
+ * Deletes the null entries from the given thought indices update and return the updated state.
+ */
+const getUpdatedState = (
+  state: State,
+  contextIndexUpdates: Index<Parent | null>,
+  thoughtIndexUpdates: Index<Lexeme | null>,
+): State => {
+  const deletedLexemeKeys = Object.keys(thoughtIndexUpdates).filter(key => !thoughtIndexUpdates[key])
+  const deletedParentKeys = Object.keys(contextIndexUpdates).filter(key => !contextIndexUpdates[key])
+
+  const updatedContextIndex = omit(
+    {
+      ...state.thoughts.contextIndex,
+      ...contextIndexUpdates,
+    },
+    [...deletedParentKeys],
+  ) as Index<Parent>
+
+  const updatedThoughtIndex = omit(
+    {
+      ...state.thoughts.thoughtIndex,
+      ...thoughtIndexUpdates,
+    },
+    [...deletedLexemeKeys],
+  ) as Index<Lexeme>
+
+  return {
+    ...state,
+    thoughts: {
+      ...state.thoughts,
+      contextIndex: {
+        ...state.thoughts.contextIndex,
+        ...updatedContextIndex,
+      },
+      thoughtIndex: {
+        ...state.thoughts.thoughtIndex,
+        ...updatedThoughtIndex,
+      },
+    },
+  }
+}
+
+/**
+ * Merges two given thoughts within same context and with same value.
+ */
+const mergeThoughts = (
```

Comment:
I'm also not clear why this is a selector rather than a reducer? It sounds as if it is meant to change state.